File: codeforces/1600~1699/1687/b.cc
#include <bits/stdc++.h>

using namespace std;

using i64 = long long;

i64 query(const string& s) {
  cout << "? " << s << endl;
  i64 L = 0;
  cin >> L;
  return L;
}

void answer(int L) {
  cout << "! " << L << endl;
}

void solve() {
  int n = 0, m = 0;
  cin >> n >> m;

  vector<i64> length(m+1);
  for (int i = 1; i <= m; ++i) {
    string s = string(m, '0');
    s[i-1] = '1';
    length[i] = query(s);
  }

  vector<pair<i64,int>> tracks;
  for (int i = 1; i <= m; ++i) {
    tracks.push_back({length[i], i});
  }

  sort(tracks.begin(), tracks.end());

  string s(m, '0');
  i64 ans = 0;
  for (int i = 0; i < m; ++i) {
    int idx = tracks[i].second - 1;
    s[idx] = '1' ;
    i64 value = query(s);
    if (value == ans + tracks[i].first) {
      ans = value;
    } else {
      s[idx] = '0';
    }
  }

  answer(ans);
}

int main(void) {
  ios::sync_with_stdio(false);
  cin.tie(0);

  solve();
  return 0;
}

File: codeforces/1600~1699/1687/b.rs
use std::io;
use std::str::FromStr;
use std::fmt::Debug;

fn read_int<T: FromStr>() -> T where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    io::stdin().read_line(&mut s).unwrap();
    s.trim().parse::<T>().expect("invalid input")
}

fn read_vec<T: FromStr>() -> Vec<T> where <T as FromStr>::Err: Debug {
    let mut s = String::new();
    io::stdin().read_line(&mut s).unwrap();
    s.split(" ").map(|x| x.trim().parse().expect("invalid input")).collect()
}

fn query(s: String) -> i64 {
    println!("? {}", s);
    read_int::<i64>()
}

fn answer(l: i64) {
    println!("! {}", l);
}

fn solve() {
    let inputs = read_vec::<usize>();
    let m = inputs[1];

    let mut s = vec![b'0'; m];
    let mut lengths = vec![0; m];
    for i in 0..m {
        s[i] = b'1';
        lengths[i] = query(String::from_utf8(s.clone()).unwrap());
        s[i] = b'0';
    }

    let mut tracks: Vec<(i64,usize)> = Vec::new();
    for i in 0..m {
        tracks.push((lengths[i], i));
    }
    tracks.sort();

    let mut ans = 0;
    for i in 0..m {
        let idx = tracks[i].1;
        s[idx] = b'1';
        let value = query(String::from_utf8(s.clone()).unwrap());
        if value == ans + tracks[i].0 {
            ans = value;
        } else {
            s[idx] = b'0';
        }
    }

    answer(ans);
}

fn main() {
    solve();
}
